README.org
#+Title: dired-rsync -- asynchronous rsync from dired

[[https://stable.melpa.org/#/dired-rsync][file:https://stable.melpa.org/packages/dired-rsync-badge.svg]]
[[https://melpa.org/#/dired-rsync][file:https://melpa.org/packages/dired-rsync-badge.svg]]
[[https://coveralls.io/repos/github/stsquad/dired-rsync][file:https://coveralls.io/repos/github/stsquad/dired-rsync/badge.svg?branch=master]]

* Introduction

This package adds the command ~dired-rsync~ which allows the user to
copy marked files in a ~dired~ buffer via rsync. This is useful,
especially for large files, because the copy happens in the background
and doesn't lock up Emacs. It is also more efficient than using tramp's
own encoding methods for moving data between systems.

Configuration is simple as you only need to bind the ~dired-rsync~
command to your preferred ~dired~ binding.

#+name: configuring-dired-rsync
#+begin_src emacs-lisp
(use-package dired-rsync
  :bind (:map dired-mode-map
              ("C-c C-r" . dired-rsync))
#+end_src

A second package in this repository provides a transient version
called ~dired-rsync-transient~. This wraps the command in a `magit`
like transient interface allowing you to tweaks the parameters for
your call.

#+name: configuring-dired-rsync-transient
#+begin_src emacs-lisp
(use-package dired-rsync-transient
  :bind (:map dired-mode-map
              ("C-c C-x" . dired-rsync-transient)))
#+end_src

* Customisation

A variable called ~dired-rsync-modeline-status~ is provided for mode
lines that will report the number of active rsync operations in
progress. The variable ~dired-rsync-job-count~ contains the number of
currently active ~dired-rsync~ jobs currently running.

For those that don't like the way ~dired-rsync~ unmarks transferred
files when it completes please see ~dired-rsync-unmark-on-completion~.

If you want to tweak the way marks are collected you can customise
~dired-rsync-source-files~ and provide your own function to return a
list of fully qualified names to include in the command.

From time to time the call to rsync may fail. dired-rsync keeps the
process buffer around for debugging and reports to the console. You
can customise ~dired-rsync-failed-hook~ with your own hook function or
select the more aggressive ~dired-rsync--pop-to-rsync-failed-buf~ to
pop straight to the buffer. Also you can customise ~dired-rsync-success-hook~
with your own hook function to provide notification for compleated transfers.

The options ~dired-rsync-command~ and ~dired-rsync-options~ are there
to modify the call to rsync but a user is unlikely to need to tweak
these.

* Technical Notes

While you can use rsync to copy files locally the main use case is
copying files to/from a remote system. The rsync tool is always run
locally as rsync needs working SSH authentication to work. If you can
access a remote machine via tramp/ssh without having to enter a
password (because ssh-agent is working) then rsync should work fine.

You can also do a remote-to-remote copy although this does involve
setting up a port forward so the first remote can access the second
remote. Also as dired-rsync needs to reverse ssh from your source
location it is best if you fully qualify the location in tramp (i.e.
user@remote:/dest/path) so it can extract the correct username to
login with. It's ugly but it works.

The test used by ~dired-rsync~ is ~tramp-tramp-file-p~ which will mean
any tramp path will attempt to be converted to an rsync path with
optional ssh transport. Obviously if your remote target doesn't also
have ssh running and remote rsync binary this will fail. The reporting
could be improved.

There have been several attempts at doing this but I found them
wanting in usability. This attempts to clean up the ideas from:

 - https://truongtx.me/tmtxt-dired-async.html
 - https://oremacs.com/2016/02/24/dired-rsync/
 - https://vxlabs.com/2018/03/30/asynchronous-rsync-with-emacs-dired-and-tramp/

in a hopefully cleaner and more idiomatic way.


NEWS
v0.7
   - truncate process buffer output to last status line
   - raise minimum Emacs version to 25.1
   - new dired-rsync-transient frontend
   - various code clean-ups
   - new customisation hook dired-rsync-success-hook
   - moved the project CI to GitHub Actions
   - now understands alternate port syntax in tramp URIs
v0.6
   - use tramp functions to decompose URIs (fix #22)
   - use username from dired URI if we have it in rsync
   - fix escaping of 's for remote-to-remote copies
   - better detection of running jobs by parsing buffers
   - increment the reverse proxy port based on active connections
   - don't set mode-line-process, leave it to user to put dired-rsync-modeline-status somewhere
v0.5
   - Big CI clean-ups (dropped emake, added Cask)
   - smarter modeline status (show R:n% when one job running)
   - make dired-rsync process dest with expand-file-name
   - we now support remote to remote with an ugly port forward hack
   - prompt the user for a passphrase if we detect stall in process
   - add some basic ert tests
   - new customisation option dired-rsync-source-files
   - new customisation hook dired-rsync-failed-hook
v0.4
   - convert dired-rsync-[command|options] to defcustom
   - added dired-rsync-unmark-on-completion option
   - re-factored to keep temp data in lexically let variable
   - and tramp path (as defined by tramp-tramp-file-p) will attempt to be converted to remote rsync
   - updates to README.org
   - A simple Travis base CI has been enabled
v0.3
   - clean-up handling of finished tasks
   - rationalise the quoting of arguments
   - automatically set mode-line-process with status
   - added LICENSE file to repo
v0.2
   - first release to MELPA


dired-rsync-ert.el
;;; dired-rsync-ert --- tests for dired-rsync
;;
;;; Commentary:
;;
;; Mostly this is for the internal functions as it is hard to test
;; remote SSH accessible resources in a test case.
;;
;;; Code:

(require 'ert)

;;
(when (and (version<= "26.1" emacs-version)
           (require 'undercover nil t))
  (undercover "dired-rsync.el"))

(require 'dired-rsync)

(ert-deftest dired-rsync-test-extract-host ()
  "Test the various extractions of host from sources."
  (should (string-equal "host"
                        (dired-rsync--extract-host-from-tramp
                         "/ssh:host:/path/to/file.txt")))
  (should (string-equal "user@host"
                        (dired-rsync--extract-host-from-tramp
                         "/ssh:user@host:/path/to/file.txt")))
  (should (string-equal "host"
                        (dired-rsync--extract-host-from-tramp
                         "/ssh:user@host:/path/to/file.txt" t))))

(ert-deftest dired-rsync-test-extract-user ()
  "Test the various extractions of user from paths."
  (should (string-equal "user"
                        (dired-rsync--extract-user-from-tramp
                         "/ssh:user@host:/path/to/file.txt")))
  (let ((tramp-default-user "wibble"))
    (should (string-equal "wibble"
                          (dired-rsync--extract-user-from-tramp
                           "/ssh:host:/path/to/file.txt")))))

(ert-deftest dired-rsync-test-extract-port ()
  "Test the various extractions of port from paths."
  (should-not (dired-rsync--extract-port-from-tramp "/path/to/file.txt"))
  (should (string-equal "1022"
                        (dired-rsync--extract-port-from-tramp
                         "/ssh:user@host#1022:/path/to/file.txt"))))

(ert-deftest dired-rsync-test-extract-path()
  "Test the various extractions of the path."
  (should (string-equal "/path/to/file.txt"
                        (car (dired-rsync--extract-paths-from-tramp
                              '("/ssh:host:/path/to/file.txt"
                                "/ssh:host:/path/to/file2.txt")))))
  (should (string-equal "/path/to/file2.txt"
                        (nth 1 (dired-rsync--extract-paths-from-tramp
                              '("/ssh:host:/path/to/file.txt"
                                "/ssh:host:/path/to/file2.txt")))))
  (should (string-equal "/path/to/file.txt"
                        (car (dired-rsync--extract-paths-from-tramp
                              '("/ssh:host:/path/to/file.txt")))))
  (should (string-equal "/path/to/pluralised\\'s.txt"
                        (car (dired-rsync--extract-paths-from-tramp
                              '("/ssh:host:/path/to/pluralised's.txt")))))
  (should (string-equal "/path/to/file.txt"
                        (car (dired-rsync--extract-paths-from-tramp
                              '("/ssh:servername|sudo:root@servername:/path/to/file.txt"))))))

(ert-deftest dired-rsync-test-quote-and-maybe-convert-from-tramp ()
  "Test quote and maybe convert from tramp defun"
  ;; test against regression of issue #26: missing username in rsync command
  (should (string-equal "username@192.168.1.1:/blat/blot/"
                        (dired-rsync--quote-and-maybe-convert-from-tramp "/scp:username@192.168.1.1:/blat/blot/")))
  (should (string-equal "192.168.1.1:/blat/blot/"
                        (dired-rsync--quote-and-maybe-convert-from-tramp "/scp:192.168.1.1:/blat/blot/"))))

(ert-deftest dired-rsync-test-remote-port()
  "Test the remote port handling."
  (should (= 50000 (dired-rsync--get-remote-port)))
  (cl-letf (((symbol-function 'dired-rsync--get-active-buffers) (lambda() '(1 2))))
   (should (= 50002 (dired-rsync--get-remote-port)))))

(ert-deftest dired-rsync-test-remote-remote-cmd ()
  "Test we generate a good remote to remote command."
  (should (string-equal
           "ssh -A -R localhost:50000:host:22 seed \"rsync -az --info=progress2 -e \\\"ssh -p 50000 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null\\\" -- a b c's user@localhost:/video\""
           (dired-rsync--remote-to-remote-cmd "seed" nil '("a" "b" "c's") "user"
                                              "host" nil "/video")))
  (should (string-equal
           "ssh -A -p 23 -R localhost:50000:host:1022 seed \"rsync -az --info=progress2 -e \\\"ssh -p 50000 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null\\\" -- a b c's user@localhost:/video\""
           (dired-rsync--remote-to-remote-cmd "seed" "23" '("a" "b" "c's") "user"
                                              "host" "1022" "/video")))
  (cl-letf (((symbol-function 'dired-rsync--get-active-buffers) (lambda() '(1 2))))
    (should (string-equal
           "ssh -A -R localhost:50002:host:22 seed \"rsync -az --info=progress2 -e \\\"ssh -p 50002 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null\\\" -- a b c's user@localhost:/video\""
           (dired-rsync--remote-to-remote-cmd "seed" nil '("a" "b" "c's") "user"
                                              "host" nil "/video")))))

;;; dired-rsync-ert.el ends here


dired-rsync-transient.el
;;; dired-rsync-transient.el --- Transient command for dired-rsync -*- lexical-binding: t -*-
;;
;; Copyright (C) 2018, 2019, 2020 Alex Bennée
;;
;; Author: Alex Bennée <alex@bennee.com>
;; Maintainer: Alex Bennée <alex@bennee.com>
;; Version: 0.6
;; Package-Requires: ((dired-rsync "0.6") (transient "0.3.0") (emacs "24.4"))
;; Homepage: https://github.com/stsquad/dired-rsync
;;
;; This file is not part of GNU Emacs.
;;
;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; This file provides a command `dired-rsync-transient', which allows
;; to alter `dired-rsync-options' temporarily with transient pop-up.
;;

;;; Code:

(require 'dired-rsync)
(require 'transient)

;;;###autoload (autoload 'dired-rsync-transient "dired-rsync-transient" nil t)
(transient-define-prefix dired-rsync-transient ()
  "Transient command for `dired-rsync'."
  :value '("-a" "-z" "--info=progress2")
  ["Arguments"
   ("-a" "archive mode; equals to -rlptgoD" "-a")
   ("-r" "recurse into directories" "-r")
   ("-z" "compress file data during the transfer" "-z")
   ("-c" "skip based on checksum, not mod-time & size" "-c" :level 6)
   ("-C" "auto-ignore files in the same way CVS does" "-C" :level 6)
   ("-u" "skip files that are newer on the receiver" "-u" :level 6)
   ("-m" "prune empty directory chains from file-list" "-m" :level 6)
   ("=d" "delete extraneous files from dest dirs" "--delete" :level 6)
   ("=s" "skip files that match in size" "--size-only" :level 6)
   ("=i" "include files matching PATTERN" "--include="
    :multi-value t :reader dired-rsync-transient--read-multiple
    :prompt "include (e.g. ‘*.pdf’ or ‘*.org, *.el’): " :level 6)
   ("=e" "exclude files matching PATTERN" "--exclude="
    :multi-value t :reader dired-rsync-transient--read-multiple
    :prompt "exclude (e.g. ‘.git’ or ‘*.bin, *.elc’): ")
   ("=o" "other rsync options" "--options="
    :prompt "Other options (e.g. -v -n): ")]
  ["Information output"
   ("-i" "output a change-summary for all updates" "-i")
   ("-h" "output numbers in a human-readable format" "-h")
   ("=I" "per-file (1) or total transfer (2) progress" "--info="
    :choices ("progress1" "progress2"))]
  ["Directory"
   ("d" "Destination" "--dest="
    :init-value (lambda (o) (oset o value (dired-dwim-target-directory)))
    :reader transient-read-directory
    :prompt "rsync to: "
    :always-read t)]
  ["Action"
   [("x" "execute" dired-rsync-transient--execute)]
   [("c" "copy command" dired-rsync-transient--copy :transient t)]]
  (interactive nil dired-mode)
  (transient-setup 'dired-rsync-transient))

(defun dired-rsync-transient--read-multiple (prompt &optional _initial-input history)
  "Read multiple values after PROMPT with optional INITIAL_INPUT and HISTORY."
  (completing-read-multiple prompt nil nil nil nil history))

(defun dired-rsync-transient--get-options (args)
  "Extract and combine rsync options from ARGS."
  (let (opts)
    (dolist (arg args)
      (cond
       ((listp arg)
        (setq opts (append
                    (mapcar (lambda (x) (format "%s'%s'" (car arg) x)) (cdr arg))
                    opts)))
       ((string-prefix-p "--options=" arg)
        (push (string-remove-prefix "--options=" arg) opts))
       ((string-prefix-p "--dest=" arg))
       (t (push arg opts))))
    (string-join (nreverse opts) " ")))

(defun dired-rsync-transient--execute (args)
  "Execute rsync command generated by transient ARGS."
  (interactive (list (transient-args transient-current-command)))
  (let ((dest (transient-arg-value "--dest=" args))
        (dired-rsync-options (dired-rsync-transient--get-options args)))
    (dired-rsync dest)))

(defun dired-rsync-transient--copy (args)
  "Copy rsync command generated by transient ARGS."
  (interactive (list (transient-args transient-current-command)))
  (let* ((dest (transient-arg-value "--dest=" args))
         (dired-rsync-options (dired-rsync-transient--get-options args))
         (cmd (dired-rsync--build-cmd (funcall dired-rsync-source-files) dest)))
    (kill-new cmd)
    (message "Command `%s' copied." cmd)))

;; don't show commands in M-x
(put 'dired-rsync-transient--copy 'completion-predicate 'ignore)
(put 'dired-rsync-transient--execute 'completion-predicate 'ignore)

(provide 'dired-rsync-transient)
;;; dired-rsync-transient.el ends here


dired-rsync.el
;;; dired-rsync.el --- Allow rsync from dired buffers -*- lexical-binding: t -*-
;;
;; Copyright (C) 2018, 2019, 2020 Alex Bennée
;;
;; Author: Alex Bennée <alex@bennee.com>
;; Maintainer: Alex Bennée <alex@bennee.com>
;; Version: 0.6
;; Package-Requires: ((s "1.12.0") (dash "2.0.0") (emacs "25.1"))
;; Homepage: https://github.com/stsquad/dired-rsync
;;
;; This file is not part of GNU Emacs.
;;
;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; dired-rsync is a command that can be run from a dired buffer to
;; copy files using rsync rather than tramps in-built mechanism.
;; This is especially useful for copying large files to/from remote
;; locations without locking up tramp.
;;
;; To use simply open a dired buffer, mark some files and invoke
;; dired-rsync.  After being prompted for a location to copy to an
;; inferior rsync process will be spawned.
;;
;; Wherever the files are selected from the rsync will always run from
;; your local machine.
;;

(require 'tramp) ; for tramp-tramp-file-p
(require 'dired-aux) ; for dired-dwim-target-directory
(require 'dash)
(require 's)
(require 'rx)

;;; Code:

;; Customisation options

(defcustom dired-rsync-command "rsync"
  "The rsync binary that we are going to use."
  :type 'string
  :group 'dired-rsync)

(defcustom dired-rsync-options "-az --info=progress2"
  "The default options for the rsync command."
  :type 'string
  :group 'dired-rsync)

(defcustom dired-rsync-unmark-on-completion t
  "Control if dired-rsync should unmark when complete."
  :type 'boolean
  :group 'dired-rsync)

(defun dired-rsync--default-fetch-marked-files ()
  "Default fetcher of marked files."
  (dired-get-marked-files nil current-prefix-arg))

(defcustom dired-rsync-source-files 'dired-rsync--default-fetch-marked-files
  "Function to collect the list of source files from dired."
  :type 'function
  :group 'dired-sync)

(defun dired-rsync--default-rsync-failed ()
  "Report rsync failure to console."
  (message "dired-rsync: failed (see %s for details)"
           (current-buffer)))

(defun dired-rsync--pop-to-rsync-failed-buf ()
  "Jump to a recently failed buffer."
  (pop-to-buffer-same-window (current-buffer)))

(defcustom dired-rsync-failed-hook '(dired-rsync--default-rsync-failed)
  "Hook run when rsync fails.
It is run in the context of the failed process buffer."
  :type 'hook
  :group 'dired-rsync)

(defcustom dired-rsync-success-hook nil
  "Hook run when rsync success."
  :type 'hook
  :group 'dired-rsync)

;; Internal variables
(defvar dired-rsync-proc-buffer-prefix "*rsync"
  "Prefix for process buffers.")

(defvar dired-rsync-modeline-status
  ""
  "A string defining current `dired-rsync' status, useful for modelines.")

(defvar dired-rsync-passphrase-stall-regex
  (rx "Enter passphrase for key")
  "A regex to detect passphrase prompts.")

(defvar dired-rsync-percent-complete-regex
  (rx (** 1 3 digit) "%")
  "A regex to extract the % complete from a file.")

(defvar dired-rsync-remote-portfwd
  "ssh -p %d -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
  "An explicit ssh command for rsync to use port forwarded proxy.
The string is treated as a format string where %d is replaced with the
results of `dired-rsync--get-remote-port'.")

;; Helpers
(defun dired-rsync--get-remote-port ()
  "Return the remote port we shall use for the reverse port-forward."
  (+ 50000 (length (dired-rsync--get-active-buffers))))

(defun dired-rsync--get-remote-portfwd ()
  (format dired-rsync-remote-portfwd (dired-rsync--get-remote-port)))

(defun dired-rsync--quote-and-maybe-convert-from-tramp (file-or-path)
  "Reformat a tramp FILE-OR-PATH to one usable for rsync."
  (if (tramp-tramp-file-p file-or-path)
      (with-parsed-tramp-file-name file-or-path tfop
        (format "%s%s:%s" (if tfop-user (format "%s@" tfop-user) "") tfop-host
                (shell-quote-argument tfop-localname)))
    (shell-quote-argument file-or-path)))

(defun dired-rsync--extract-host-from-tramp (file-or-path &optional split-user)
  "Extract the tramp host part of FILE-OR-PATH.

It SPLIT-USER is set we remove the user@ part as well.  We assume
hosts don't need quoting."
  (with-parsed-tramp-file-name file-or-path tfop
    (if (or split-user (not tfop-user))
        tfop-host
      (format "%s@%s" tfop-user tfop-host))))

;; This is tricky for remote-to-remote because we may have an implied
;; user from the local config which isn't available on the remote
;; .ssh/config
(defun dired-rsync--extract-user-from-tramp (file-or-path)
  "Extract the username part of a tramp FILE-OR-PATH."
  (with-parsed-tramp-file-name file-or-path tfop
    (or tfop-user
        ; somehow extract .ssh/config user for tfop-host
        (getenv "USER"))))

(defun dired-rsync--extract-port-from-tramp (file-or-path)
  "Extract the port part of a tramp FILE-OR-PATH."
  (when (tramp-tramp-file-p file-or-path)
    (with-parsed-tramp-file-name file-or-path tfop
      tfop-port)))

(defun dired-rsync--extract-paths-from-tramp (files)
  "Extract the path part of a tramp FILES and quote it."
  (--map
   (with-parsed-tramp-file-name it tfop
     (shell-quote-argument tfop-localname))
   files))


;; Count active buffers
(defun dired-rsync--get-proc-buffers ()
  "Return all dired-rsync process buffers."
  (--filter
   (and (s-starts-with? dired-rsync-proc-buffer-prefix (buffer-name it))
        (get-buffer-process it))
   (buffer-list)))

(defun dired-rsync--get-active-buffers ()
  "Return all currently active process buffers"
  (--filter
   (process-live-p (get-buffer-process it))
   (dired-rsync--get-proc-buffers)))


;; Update status with count/speed
(defun dired-rsync--update-modeline (&optional err ind)
  "Update the modeline, optionally with `ERR' or `IND'.

`ERR' is set this indicates a problem, otherwise `IND' is an
alternative indication (such as a percentage completion).  If
neither is set we simply display the current number of jobs."
  (force-mode-line-update)
  (let ((job-count (length (dired-rsync--get-active-buffers))))
    (setq dired-rsync-modeline-status
          (cond
           ;; error has occurred
           (err (propertize
                 (format " R:%d %s!!" job-count err)
                 'font-lock-face '(:foreground "red")))
           ;; we still have jobs but no error
           ((> job-count 1)
            (format " R:%d" job-count))
           ((> job-count 0)
            (format " R:%s" (or ind job-count)))
           ;; Any stale?
           ((dired-rsync--get-proc-buffers)
            (propertize
                 " R:hung :-("
                 'font-lock-face '(:foreground "red")))
           ;; nothing going on
           (t nil)))))

;;
;; Running rsync: We need to take care of a couple of things here. We
;; need to ensure we run from the local host as you shouldn't expect
;; the remote target to be as aware of the ssh shortcuts home as from
;; the local system out (.ssh/config). We also want to track when it
;; is finished so we can inform the user the copy is complete.
;;

(defun dired-rsync--sentinel(proc desc details)
  "Process sentinel for rsync processes.
This gets called whenever the inferior `PROC' changes state as
  described by `DESC'.  `DETAILS' provides access to additional
  information such as the locate of the dired-buffer."
  (let ((proc-buf (process-buffer proc)))
    (when (s-starts-with-p "finished" desc)
      ;; clean-up finished tasks
      (let ((dired-buf (plist-get details ':dired-buffer)))
        (when (and dired-rsync-unmark-on-completion
                   (buffer-live-p dired-buf))
          (with-current-buffer dired-buf
            (dired-unmark-all-marks)))
        (kill-buffer proc-buf))
      (run-hooks 'dired-rsync-success-hook))
    (dired-rsync--update-modeline)
    ;; If we still have a process buffer things didn't end well
    (when (and (not (process-live-p proc))
               (buffer-name proc-buf))
      (with-current-buffer proc-buf
        (run-hooks 'dired-rsync-failed-hook)))))


(defun dired-rsync--filter (proc string)
  "`PROC' rsync process filter, insert `STRING' into buffer.

This gets called with string whenever there is new data to
display in the process buffer.  We scan the string to extract useful
information and update the dired-rsync-modeline-status."

  ;; scan the new string
  (let ((err nil) (indicator nil))
    ;; Grab % complete string
    (when (string-match dired-rsync-percent-complete-regex string)
      (setq indicator (match-string 0 string)))
    ;; check for prompt
    (when (string-match dired-rsync-passphrase-stall-regex string)
      (process-send-string proc (concat (read-passwd string) "\n")))
    ;; update if anything to report
    (when (or err indicator)
      (dired-rsync--update-modeline err indicator)))

  ;; update the process buffer (we could just drop?)
  (let ((old-process-mark (process-mark proc)))
    ;; do the normal buffer text insertion
    (when (buffer-live-p (process-buffer proc))
      (with-current-buffer (process-buffer proc)
        (let ((moving (= (point) old-process-mark)))
          (save-excursion
            ;; Insert the text, advancing the process marker.
            (goto-char old-process-mark)
            (insert string)
            (set-marker (process-mark proc) (point))
            ;; Delete old text upto the newline
            (goto-char (point-max))
            (when (search-backward "\r")
              (delete-region (point-min) (+ 1 (match-beginning 0)))))
          (if moving
              (goto-char (process-mark proc))))))))


(defun dired-rsync--do-run (command details)
  "Run rsync COMMAND in a unique buffer, passing DETAILS to sentinel."
  (apply #'make-process
	 (append (list :name "*rsync*"
		       :buffer (format "%s @ %s"
				       dired-rsync-proc-buffer-prefix
				       (current-time-string))
		       :command (list shell-file-name
				      shell-command-switch
				      command)
		       :sentinel (lambda (proc desc)
				   (dired-rsync--sentinel proc desc details))
		       :filter (lambda (proc string)
				 (dired-rsync--filter proc string)))))
  (dired-rsync--update-modeline))

(defun dired-rsync--remote-to-from-local-cmd (sfiles dest)
  "Construct a rsync command for SFILES to DEST copy.

This handles both remote to local or local to remote copy.
Fortunately both forms are broadly the same."
  (let ((src-files
         (-map 'dired-rsync--quote-and-maybe-convert-from-tramp sfiles))
        (final-dest (dired-rsync--quote-and-maybe-convert-from-tramp dest))
        (ssh-port (-some #'dired-rsync--extract-port-from-tramp
                         (append (list dest) sfiles))))
    (s-join " "
            (-flatten
             (list dired-rsync-command
                   dired-rsync-options
		   (when ssh-port (format "-e 'ssh -p %s'" ssh-port))
                   "--"
                   src-files
                   final-dest)))))

;; ref: https://unix.stackexchange.com/questions/183504/how-to-rsync-files-between-two-remotes
(defun dired-rsync--remote-to-remote-cmd (shost sport sfiles duser dhost dport dpath)
  "Construct and trigger an rsync run for remote copy.
The source SHOST and SFILES to remote DUSER @ DHOST to DPATH.

rsync doesn't support this mode of operation but we can fake it by
providing a port forward from the source host which we pass onto the
destination.  This requires ssh'ing to the source and running the rsync
there."
  (s-join " " (-flatten
               (list "ssh" "-A" (when sport (format "-p %s" sport))
                     "-R" (format "localhost:%d:%s:%s"
                                  (dired-rsync--get-remote-port) dhost
                                  (or dport "22"))
                     shost
                     (format
                      "\"%s %s -e \\\"%s\\\" -- %s %s@localhost:%s\""
                      dired-rsync-command
                      dired-rsync-options
                      (dired-rsync--get-remote-portfwd)
                      (s-join " " sfiles)
                      duser
                      dpath)))))

(defun dired-rsync--build-cmd (sfiles dest)
  "Construct a rsync command for SFILES to DEST copy."
  (if (and (tramp-tramp-file-p dest)
           (tramp-tramp-file-p (-first-item sfiles)))
      (let ((shost (dired-rsync--extract-host-from-tramp (-first-item sfiles)))
            (sport (dired-rsync--extract-port-from-tramp (-first-item sfiles)))
            (src-files (dired-rsync--extract-paths-from-tramp sfiles))
            (dhost (dired-rsync--extract-host-from-tramp dest t))
            (dport (dired-rsync--extract-port-from-tramp dest))
            (duser (dired-rsync--extract-user-from-tramp dest))
            (dpath (-first-item (dired-rsync--extract-paths-from-tramp (list dest)))))
        (dired-rsync--remote-to-remote-cmd shost sport src-files
                                           duser dhost dport dpath))
    (dired-rsync--remote-to-from-local-cmd sfiles dest)))

;;;###autoload
(defun dired-rsync (dest)
  "Asynchronously copy files in dired to `DEST' using rsync.

`DEST' can be a relative filename and will be processed by
`expand-file-name' before being passed to the rsync command.

This function runs the copy asynchronously so Emacs won't block whilst
the copy is running.  It also handles both source and destinations on
ssh/scp tramp connections."
  ;; Interactively grab dest if not called with
  (interactive
   (list (read-file-name "rsync to: " (dired-dwim-target-directory)
                         nil nil nil 'file-directory-p)))

  (setq dest (expand-file-name dest))

  (let* ((sfiles (funcall dired-rsync-source-files))
         (cmd (dired-rsync--build-cmd sfiles dest)))
    (dired-rsync--do-run cmd
                         (list :marked-files sfiles
                               :dired-buffer (current-buffer)))))

(provide 'dired-rsync)
;;; dired-rsync.el ends here
